sdk/go/pulumi/types.go

```diff
@@ -1017,6 +1017,31 @@ func (ResourceOutput) ElementType() reflect.Type {
 	return reflect.TypeOf((*Resource)(nil)).Elem()
 }
 
+func (o ResourceOutput) ToResourceOutput() ResourceOutput {
+	return o
+}
+
+func (o ResourceOutput) ToResourceOutputWithContext(ctx context.Context) ResourceOutput {
+	return o
+}
+
+// An Input type carrying Resource values.
+//
+// Unfortunately `Resource` values do not implement `ResourceInput` in
```

Comment:
I'm running into a lot of difficulty trying to make `Resource` values compatible with `ResourceInput`. I'm starting to think the problem is with "deep reflect" combinators that try to traverse values as trees, find all Inputs and await them.. 

I also worry about these combinators attempting to await extraneous/outputs inputs on Resource properties and try to count them toward dependencies. If we have `ResourceInput` carrying a `r: Resource` , when we count the dependency closure, we want dependencies of the `Input[]` and the `r.urn`, and we probably don't want the deps of `r.Output1` etc.

Reply:
> I'm running into a lot of difficulty trying to make Resource values compatible with ResourceInput. I'm starting to think the problem is with "deep reflect" combinators that try to traverse values as trees, find all Inputs and await them..

Can you elaborate on the issues you're running into?

> we probably don't want the deps of r.Output1 etc.

FWIW, in practice the deps of `r.Output1` for a custom resource are `[]Resource{r}`. But I think that you're right in general.

Reply:
Very specifically this: https://github.com/pulumi/pulumi/pull/7394

Substitute `Example` with `Resource` to make this practical. When interfaces travel through input/output machinery something in the generic/reflecty code gets off the rails, and I tried but failed to fix that.

Reply:
With the change in #7394, are you suggesting we'd change `Resource` to implement `ResourceInput`?  Or would we make `ResourceState` implement `ResourceInput`?

Is that change necessary for this PR or can it be deferred?

Reply:
I'm struggling to make the suggestion more specific. For the users' sake, I felt it might be "nice" to be able to pass resources where resource inputs are expected, e.g.:

```
   func f(x ResourceInput)
   res := NewFooResource("my-res1")
   f(res)

   # vs
   
   f(NewResourceInput(res))
```

Initially I thought this would need:

```
   type Resource interface {
        ...
	ResourceInput
   }
```

But your comment makes me realize specific resource types can implement ResourceInput without making Resource implement ResourceInput. So far so good.

It was tempting to make some changes to `ResourceState` so that *all* structs embedding ResourceState automatically start implementing ResourceInput. This however is not possible AFAIK, because:

```
type Input interface {
	ElementType() reflect.Type
}
type ResourceInput interface {
        Input
}
```

So `ResourceState` cannot implement `ElementType()` since it does not know the struct it is being embedded in, and AFAIK it cannot capture `self` referring to that outer struct. This is somewhat unlike classical inheritance.

Reply:
Now that I think of it, we could follow up with a codegen change though to make provider resources implement ResourceInput.

Reply:
These are often going to show up as outputs from multi-lang components like [this](https://github.com/pulumi/pulumi-eks/blob/86861e6cb6091e97d08828a444bb70087a8701f3/sdk/go/eks/managedNodeGroup.go#L24):

```go
type ManagedNodeGroup struct {
	pulumi.ResourceState

	// The AWS managed node group.
	NodeGroup eks.NodeGroupOutput `pulumi:"nodeGroup"`
}
```

If I have an instance of the `ManagedNodeGroup` component, and I want to pass its `NodeGroup` output (typed as `github.com/pulumi/pulumi-aws/sdk/v4/go/aws/eks.NodeGroupOutput`) to `DependsOnInputs`, what would that look like with just the changes in this PR?

Would I have to use an `ApplyT` to do the conversion like the following?

```go
ng, err := eks.NewManagedNodeGroup(ctx, "ng", &eks.ManagedNodeGroupArgs{...})
pulumi.DependsOnInputs(pulumi.ResourceArray{
	ng.NodeGroup.ApplyT(func(v pulumi.Resource) pulumi.Resource {
		return v
	}).(pulumi.ResourceOutput),
})
```

Reply:
I filed this as https://github.com/pulumi/pulumi/issues/7750